isabelle/Core/AssetsPreservation.thy

```diff
@@ -0,0 +1,1195 @@
+
+    moreover have "assetsInInputs inputs = 0" 
+      using calculation by simp
+
+    moreover have "outPayments = payments @ rPayments"  
+                  "outState = rState" 
+      using calculation by auto
+
+    ultimately show ?thesis
+      by (metis Groups.group_cancel.add2 add_cancel_left_right assetsInPayments_append)
+   
+  next
+    case (Cons inputHead inputTail)
+    
+    show ?thesis 
+    proof (cases "applyInput env rState inputHead rCont")
+      case (Applied applyWarn applyState applyCont)
+      moreover note contractIsReduced preservedReducedAssets applyAllLoopInduction Cons
+
+      moreover have "validAndPositive_state rState"
+        using calculation reduceContractUntilQuiescent_preserves_validAndPositive_state
+        by meson 
+
+      moreover have "validAndPositive_state applyState"
+        using calculation applyInput_preserves_preserves_validAndPositive_state  
+        by meson
+
+      moreover have "applyAllLoop 
+                  True env applyState applyCont inputTail 
+                  (warnings @ (convertReduceWarnings rWarns) 
+                            @ (convertApplyWarning applyWarn)) 
+                  (payments @ rPayments)
+              = ApplyAllSuccess outContractChanged outWarnings outPayments outState cont"
+        using calculation by auto        
+      
+      moreover have  
+         "assetsInState applyState + assetsInInputs inputTail + assetsInPayments (payments @ rPayments) 
+          = assetsInState outState + assetsInPayments outPayments"
+        using calculation by blast
+
+      moreover have "assetsInState rState + assetsInInput inputHead = assetsInState applyState" 
+        using calculation by (meson applyInput_preserves_assets)        
+
+      moreover have "assetsInInputs inputs = assetsInInput inputHead + assetsInInputs inputTail"
+        using Cons by auto
+
+      ultimately show ?thesis   
+        by (metis Groups.ab_semigroup_add_class.add.commute Groups.group_cancel.add2 assetsInPayments_append)
+  
+    next
+      case ApplyNoMatchError
+      with contractIsReduced Cons applyAllLoopInduction show ?thesis by simp
+    qed    
+  qed
+qed
+
+theorem applyAllInputs_preserves_assets : 
+  assumes "validAndPositive_state inState"
+      and "applyAllInputs env inState inContract inputs' 
+          = ApplyAllSuccess outContractChanged outWarnings outPayments outState cont"
+    shows "assetsInState inState + assetsInInputs inputs' 
+          = assetsInState outState + assetsInPayments outPayments"
+  using assms applyAllLoop_preserves_assets by fastforce
+
+subsection "Compute transaction"
+
+
+theorem computeTransaction_preserves_assets :
+  assumes "validAndPositive_state state"
+      and "computeTransaction tx state contract = TransactionOutput out"
+    shows "assetsInState state + assetsInTransaction tx 
+          = assetsInState (txOutState out) + assetsInPayments (txOutPayments out)" 
```

Comment:
Almost there!
If you used pattern matching in `out` then it would look more like the other theorems

Reply:
I tried using pattern matching here but it complicates the playTrace proof (which is already complicated 😅)

Reply:
Huh, that is odd, anyway, it is fine as it is